diff -aur util-linux-2.21.1-orig/text-utils/hexdump.h util-linux-2.21.1/text-utils/hexdump.h
--- util-linux-2.21.1-orig/text-utils/hexdump.h	2012-03-29 13:20:10.781647935 +0000
+++ util-linux-2.21.1/text-utils/hexdump.h	2012-05-13 15:22:11.166096671 +0000
@@ -86,7 +86,7 @@
 void rewrite(FS *);
 void addfile(char *);
 void display(void);
-void __attribute__((__noreturn__)) usage(FILE *out);
+void __attribute__((__noreturn__)) usage(FILE *out, char *name);
 void conv_c(PR *, u_char *);
 void conv_u(PR *, u_char *);
 int  next(char **);
diff -aur util-linux-2.21.1-orig/text-utils/hexsyntax.c util-linux-2.21.1/text-utils/hexsyntax.c
--- util-linux-2.21.1-orig/text-utils/hexsyntax.c	2012-03-29 13:20:10.781647935 +0000
+++ util-linux-2.21.1/text-utils/hexsyntax.c	2012-05-13 16:03:23.806057592 +0000
@@ -112,12 +112,12 @@
 			break;
 		case 'V':
 			printf(_("%s from %s\n"),
-					program_invocation_short_name,
+					argv[0],
 					PACKAGE_STRING);
 			exit(EXIT_SUCCESS);
 			break;
 		default:
-			usage(stderr);
+			usage(stderr, argv[0]);
 		}
 	}
 
@@ -129,11 +129,11 @@
 	*argvp += optind;
 }
 
-void __attribute__((__noreturn__)) usage(FILE *out)
+void __attribute__((__noreturn__)) usage(FILE *out, char *name)
 {
 	fprintf(out, _("\nUsage:\n"
 		       " %s [options] file...\n"),
-		       program_invocation_short_name);
+		       name);
 	fprintf(out, _(
 		       "\nOptions:\n"
 		       " -b              one-byte octal display\n"
